docs/proposals/20190708-MachineConfig.md

```diff
@@ -0,0 +1,223 @@
+        cidr: x.x.x.x/32
+      - interface: eth0
+        dhcp: true
+install:
+  wipe: false
+  force: true
+  boot:
+    device: /dev/sda
+    size: 1024000000
+    kernel: http://139.178.69.21:8080/assets/talos/release/v0.1.0-beta.0/vmlinuz
+    initramfs: http://139.178.69.21:8080/assets/talos/release/v0.1.0-beta.0/initramfs.xz
+  root:
+    device: /dev/sda
+    size: 2048000000
+    rootfs: http://139.178.69.21:8080/assets/talos/release/v0.1.0-beta.0/rootfs.tar.gz
+  data:
+    device: /dev/sda
+    size: 4096000000
+```
+
+## Proposal
+
+There are quite a few fields that are similar among all node types, while the main difference between them is the kubeadm templates. Given all the proper info, we should be able to generate the kubeadm configuration on the fly depending on node type and abstract the need to know all the different options away. I'm proposing a machine configuration that looks like this:
+
+```
+#!talos
+## We'll implement a versioning strategy and 
+## respect it when parsing the config
+version: "v0.1.0"
+
+## All configs that are specific to this machine only
+## go in machine key
+machine:
+  type: init, controlplane, worker
+  kubelet:
+    image: gcr.io/google-containers/hyperkube:v1.15.2
+    extraArgs: {}
+  network:
+    hostname: "wtf.m8.cluster"
+    interfaces:
+      ##MAC AND DEVICE ARE MUTUALLY EXCLUSIVE
+      - mac: aa:bb:cc:ee:ff:00
+        addresses:
+          - "10.254.10.11/24"
+          - "192.168.251.11/24"
+          - "2001:db8:1000::54/64"
+        mtu: 1450
+        type: static/dhcp
+      - device: eth1
+        addresses:
+          - "55.44.11.23/29"
+        ## An example of how we might provide routes for a given interface
+        routes:
+          - via: 192.168.24.1
+            prefix: 0.0.0.0/0
+            gateway: true
+            metric: 100
+          - via: 2001:db8::0:2bd9
+            prefix: 2001:db8::10:0/64
+            gateway: false
+            metric: 10
+      - device: bond0
+        subdevices:
+          - eth0
+          - eth1
+  ## This is currently just an idea of how another talos "service" 
+  ## might get configured for the machine
+  ntp:
+    servers:
+      - server1
+  ca:
+    crt: "{{ .Certs.OsCert }}"
+    key: "{{ .Certs.OsKey }}"
+  token: abc.1234
+  ## We've removed the size of the "data" partition and renamed it to "ephemeral". Also removed the root partition
+  install:
+    wipe: false
+    force: true
+    boot:
+      device: /dev/sda
+      size: 1024000000
+      kernel: https://github.com/talos-systems/talos/releases/download/v0.2.0-alpha.3/vmlinuz
+      initramfs: https://github.com/talos-systems/talos/releases/download/v0.2.0-alpha.3/initramfs.xz
+    ephemeral:
+      device: /dev/sda
+
+## All cluster-wide configs go here. These are generally things that get 
+## injected to kubeadm, or affect the behavior of the kubernetes control 
+## plane as a whole.
+cluster:
+  clusterName: {{ .ClusterName }}
+  controlPlane:
+    ips: 
```

Comment:
We should also have some mechanism by which we can specify a DNS name or load balancer IP for clients to use.  Perhaps `canonicalAddress` or `advertisedAddress`?  This is important at a minimum for TLS certs (subject alternative names), but also for more advanced network configurations where some internal components need to be able to bootstrap before the service node is available.

Reply:
Perhaps that should be under `apiServer`?  Or better, `clusterAddress` next to `clusterName`.